### PR TITLE
HTTP2 scenario coverage

### LINKED ISSUE #7161: HTTP2 scenario coverage
This is for tracking notes about HTTP/2 related features in 2.2. Most of the work will be related to Kestrel, but HttpSys and IIS In-Proc will also get a few notes.

HTTP/2 support is limited by TLS ALPN availability on various platforms.

Kestrel
- OS
  - Windows 8.1+
  - Linux W/OpenSSL 1.0.2 (e.g. Ubuntu 16.04)
  - Mac - Not yet supported - https://github.com/dotnet/corefx/issues/30492
- Framework
  - NetCoreApp 2.2 or later
  - .NET Framework - Not yet supported

Not ASP.NET Core version specific, only constrained by the OS:

HttpSysServer
- Win10+
- Not framework specific

IIS In proc
- Win10+
- .NET Core (in proc restriction)

IIS Out of proc
- Win10+
- The front end can use HTTP/2, but the reverse proxy to Kestrel will continue to use HTTP/1.1.
- Not framework specific

### PR DESCRIPTION
**Pull request by guardrex**
Fixes #7161 

[Internal Review Topic (Servers landing page)](https://review.docs.microsoft.com/en-us/aspnet/core/fundamentals/servers/index?view=aspnetcore-1.0&branch=pr-en-us-8519)
[Internal Review Topic (Kestrel)](https://review.docs.microsoft.com/en-us/aspnet/core/fundamentals/servers/kestrel?view=aspnetcore-1.0&branch=pr-en-us-8519)
[Internal Review Topic (HTTP.sys)](https://review.docs.microsoft.com/en-us/aspnet/core/fundamentals/servers/httpsys?view=aspnetcore-2.0&branch=pr-en-us-8519)
[Internal Review Topic (IIS)](https://review.docs.microsoft.com/en-us/aspnet/core/host-and-deploy/iis/index?view=aspnetcore-1.0&branch=pr-en-us-8519)
[Internal Review Topic (Linux Nginx)](https://review.docs.microsoft.com/en-us/aspnet/core/host-and-deploy/linux-nginx?view=aspnetcore-1.0&branch=pr-en-us-8519)
[Internal Review Topic (Linux Apache)](https://review.docs.microsoft.com/en-us/aspnet/core/host-and-deploy/linux-apache?view=aspnetcore-1.0&branch=pr-en-us-8519)

No doubt in need of corrections and additions. I still had to guess on a few items, but it should be easy to sort these spots out on review.

Opened **Add API links to HTTP/2 content** #8518 to update the API references to links.

### CODE REVIEW

File: aspnetcore/host-and-deploy/linux-apache.md [hunk 1]
```diff
@@ -22,6 +22,21 @@ Using this guide, learn how to set up [Apache](https://httpd.apache.org/) as a r
    1. Select and follow the instructions for CentOS/Oracle.
 1. An existing ASP.NET Core app.
 
+::: moniker range=">= aspnetcore-2.2"
+
+## HTTP/2 support
+
+[HTTP/2](https://httpwg.org/specs/rfc7540.html) is available for ASP.NET Core apps running directly on the [Kestrel server](xref:fundamentals/servers/kestrel#http2-support) if the following base requirements are met:
```
Comment (Tratcher) [inline aspnetcore/host-and-deploy/linux-apache.md]: This is a bit odd in the Apache doc. Does Apache support forwarding http/2?

File: aspnetcore/host-and-deploy/linux-nginx.md [hunk 2]
```diff
@@ -34,6 +34,21 @@ This guide:
    1. Select and follow the instructions for Ubuntu that match the Ubuntu version of the server.
 1. An existing ASP.NET Core app.
 
+::: moniker range=">= aspnetcore-2.2"
+
+## HTTP/2 support
+
+[HTTP/2](https://httpwg.org/specs/rfc7540.html) is available for ASP.NET Core apps running directly on the [Kestrel server](xref:fundamentals/servers/kestrel#http2-support) if the following base requirements are met:
```
Comment (Tratcher) [inline aspnetcore/host-and-deploy/linux-nginx.md]: Same as apache

File: aspnetcore/fundamentals/servers/kestrel.md [hunk 3]
```diff
@@ -698,6 +781,129 @@ When using IIS, the URL bindings for IIS override bindings set by `UseUrls`. For
 
 ::: moniker-end
 
+::: moniker range=">= aspnetcore-2.2"
+
+**ListenOptions.Protocols**
+
+The `Protocols` property establishes the HTTP protocols (`HttpProtocols`) enabled on a connection endpoint or for the server. Assign a value to the `Protocols` property from the `HttpProtocols` enum.
+
+| `HttpProtocols` enum value | Connection protocol permitted |
```
[the remark this answers is not in the document]
Comment (guardrex) [reply aspnetcore/fundamentals/servers/kestrel.md]: Let me know if you have a link for "Prior Knowledge mode."

[the remark this answers is not in the document]
Comment (Tratcher) [reply aspnetcore/fundamentals/servers/kestrel.md]: https://tools.ietf.org/html/rfc7540#section-3.4

File: aspnetcore/fundamentals/servers/index.md [hunk 4]
```diff
@@ -102,10 +106,51 @@ When using [Visual Studio](https://www.visualstudio.com/vs/), [Visual Studio for
...
+[HTTP/2](https://httpwg.org/specs/rfc7540.html) is supported with ASP.NET Core in the following deployment scenarios:
+
+::: moniker range=">= aspnetcore-2.2"
+
+* [Kestrel](xref:fundamentals/servers/kestrel#http2-support)
+  * Operating system
+    * Windows Server 2012 R2/Windows 8.1 or later
+    * Linux with OpenSSL 1.0.2 or later (for example, Ubuntu 16.04 or later)
+    * HTTP/2 will be supported on macOS in a future release.
+  * Target framework: ASP.NET Core 2.2 or later
```
Comment (Tratcher) [inline aspnetcore/fundamentals/servers/index.md]: .NET Core 2.2, not ASP.NET Core 2.2. .NET 4.x doesn't support ALPN or HTTP/2.

File: aspnetcore/fundamentals/servers/index.md [hunk 5]
```diff
@@ -102,10 +106,51 @@ When using [Visual Studio](https://www.visualstudio.com/vs/), [Visual Studio for
...
+* [HTTP.sys](xref:fundamentals/servers/httpsys#http2-support)
+  * Windows Server 2016/Windows 10 or later
+  * Target framework: Not applicable to HTTP.sys deployments.
+* [IIS (in-process)](xref:host-and-deploy/iis/index#http2-support)
+  * Windows Server 2016/Windows 10 or later; IIS 10 or later
+  * Target framework: ASP.NET Core 2.2 or later
```
Comment (Tratcher) [inline aspnetcore/fundamentals/servers/index.md]: .NET Core 2.2, no .NET 4.x support for in-proc.

File: aspnetcore/fundamentals/servers/kestrel.md [hunk 6]
```diff
@@ -15,14 +15,50 @@ Kestrel is a cross-platform [web server for ASP.NET Core](xref:fundamentals/serv
...
+::: moniker range=">= aspnetcore-2.2"
+
+## HTTP/2 support
+
+[HTTP/2](https://httpwg.org/specs/rfc7540.html) is available for ASP.NET Core apps if the following base requirements are met:
+
+* Operating system&dagger;
+  * Windows Server 2012 R2/Windows 8.1 or later
+  * Linux with OpenSSL 1.0.2 or later (for example, Ubuntu 16.04 or later)
+* Target framework: ASP.NET Core 2.2 or later
```
Comment (Tratcher) [inline aspnetcore/fundamentals/servers/kestrel.md]: .NET Core 2.2

File: aspnetcore/fundamentals/servers/kestrel.md [hunk 7]
```diff
@@ -698,6 +781,129 @@ When using IIS, the URL bindings for IIS override bindings set by `UseUrls`. For
...
+::: moniker range=">= aspnetcore-2.2"
+
+**ListenOptions.Protocols**
+
+The `Protocols` property establishes the HTTP protocols (`HttpProtocols`) enabled on a connection endpoint or for the server. Assign a value to the `Protocols` property from the `HttpProtocols` enum.
+
+| `HttpProtocols` enum value | Connection protocol permitted |
+| -------------------------- | ----------------------------- |
+| `Http1`                    | HTTP/1.1 only. Can be used with or without TLS. |
+| `Http2`                    | HTTP/2 only. Primarily used with TLS 1.2 or later. May be used without TLS 1.2 or later only if the client supports a [Prior Knowledge mode](https://tools.ietf.org/html/rfc7540#section-3.4). |
```
Comment (Tratcher) [inline aspnetcore/fundamentals/servers/kestrel.md]: ~~1.2 or later~~ is distracting here. It's not about the version, it's a bout TLS vs clear text.

File: aspnetcore/fundamentals/servers/kestrel.md [hunk 8]
```diff
@@ -698,6 +781,129 @@ When using IIS, the URL bindings for IIS override bindings set by `UseUrls`. For
...
+| `Http1AndHttp2`            | HTTP/1.1 and HTTP/2. Requires a TLS 1.2 or later and [Application-Layer Protocol Negotiation (ALPN)](https://tools.ietf.org/html/rfc7301#section-3) connection to negotiate HTTP/2; otherwise, the connection defaults to HTTP/1.1. |
```
Comment (Tratcher) [inline aspnetcore/fundamentals/servers/kestrel.md]: Same about the version here.

File: aspnetcore/host-and-deploy/iis/index.md [hunk 9]
```diff
@@ -20,11 +20,43 @@ The following operating systems are supported:
...
 
+## HTTP/2 support
+
+::: moniker range=">= aspnetcore-2.2"
+
+[HTTP/2](https://httpwg.org/specs/rfc7540.html) is supported with ASP.NET Core in the following IIS deployment scenarios:
+
+* In-process
+  * Windows Server 2016/Windows 10 or later; IIS 10 or later
+  * Target framework: ASP.NET Core 2.2 or later
```
Comment (Tratcher) [inline aspnetcore/host-and-deploy/iis/index.md]: .NET Core 2.2

File: aspnetcore/host-and-deploy/iis/index.md [hunk 10]
```diff
@@ -20,11 +20,43 @@ The following operating systems are supported:
...
+
+::: moniker-end
+
+::: moniker range="< aspnetcore-2.2"
+
+[HTTP/2](https://httpwg.org/specs/rfc7540.html) is supported for out-of-process deployments that meet the following base requirements:
+
+* Windows Server 2016/Windows 10 or later; IIS 10 or later
+* Edge connections use HTTP/2, but the reverse proxy connection to the [Kestrel server](xref:fundamentals/servers/kestrel) uses HTTP/1.1.
+* Target framework: Not applicable to out-of-process deployments, since the HTTP/2 connection is handled entirely by IIS.
```
Comment (Tratcher) [inline aspnetcore/host-and-deploy/iis/index.md]: * TLS 1.2 or later connection

File: aspnetcore/host-and-deploy/iis/index.md [hunk 11]
```diff
@@ -20,11 +20,43 @@ The following operating systems are supported:
...
+
+::: moniker-end
+
+If an HTTP/2 connection is established, [HttpRequest.Protocol](xref:Microsoft.AspNetCore.Http.HttpRequest.Protocol*) reports `HTTP/2`.
```
Comment (Tratcher) [inline aspnetcore/host-and-deploy/iis/index.md]: HttpRequest.Protocol is only set to Http/2 for in proc, not out of proc.

File: aspnetcore/fundamentals/servers/index.md (hunk 4 above)
Comment (guardrex) [reply aspnetcore/fundamentals/servers/index.md]: Should it say "Shared framework: ..." since we don't want to confuse TFMs with this remark?

Comment (Tratcher) [reply aspnetcore/fundamentals/servers/index.md]: TFM is more applicable. In-proc and ALPN both have dependencies in .NET Core 2.2 and don't support .NET 4.x.

File: aspnetcore/fundamentals/servers/kestrel.md [hunk 12]
```diff
@@ -253,6 +289,49 @@ You can configure the rates per request in middleware:
...
 ::: moniker-end
 
+::: moniker range=">= aspnetcore-2.2"
+
+**MaxStreamsPerConnection**
+
+`Http2.MaxStreamsPerConnection` limits the number of concurrent request streams per HTTP/2 connection. Excess streams are refused.
+
+```csharp
+.ConfigureKestrel((context, options) =>
```
Comment (scottaddie) [inline aspnetcore/fundamentals/servers/kestrel.md]: Indicate where this code belongs.

File: aspnetcore/fundamentals/servers/kestrel.md [hunk 13]
```diff
@@ -698,6 +781,129 @@ When using IIS, the URL bindings for IIS override bindings set by `UseUrls`. For
...
+  * Elliptic curve Diffie-Hellman (ECDHE) &lbrack;[RFC4492](https://www.ietf.org/rfc/rfc4492.txt)&rbrack; &ndash; 224 bits minimum
+  * Finite field Diffie-Hellman (DHE) &lbrack;`TLS12`&rbrack; &ndash; 2048 bits minimum
+* Cipher suite not blacklisted
+
+`TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256` &lbrack;`TLS-ECDHE`&rbrack; with the P-256 elliptic curve &lbrack;`FIPS186`&rbrack; is supported by default.
+
+The following example permits HTTP/1.1 and HTTP/2 connections on port 8000. Connections are secured by TLS with a supplied certificate:
+
+```csharp
+options.Listen(IPAddress.Any, 8000, listenOptions =>
```
Comment (scottaddie) [inline aspnetcore/fundamentals/servers/kestrel.md]: Indicate where this code belongs.

File: aspnetcore/fundamentals/servers/kestrel.md (hunk 12 above)
Comment (guardrex) [reply aspnetcore/fundamentals/servers/kestrel.md]: I have the `ConfigureKestrel` call shown directly under the `Program.Main` (near where the topic calls out `UseKestrel` with options, too) much earlier in the topic (up around Lines 114-125). The next 11 appearances of `ConfigureKestrel` follow this format, too, to minimize the size of the examples.

I originally noted more work for this in the 2.2 master tracking issue: #7180. I've just opened a proper issue to update the sample at #8571.

**Comment by guardrex** [comment]
@scottaddie kk --- that might&dagger; get every one, including `UseKestrel`, in context.

&dagger;This topic has become a 🐻 to work on.

IIRC there will be overlap between 1.x and 3.0, as I think 1.x probably depreciates *after* 3.0 RTM.

I pray to the :pray: *server gods* :pray: for file versioning by then.

**Comment by guardrex** [comment]
ty @scottaddie 

@Tratcher U ok with the latest updates?

**Comment by Tratcher** [comment]
:shipit:

**Pull request merged**